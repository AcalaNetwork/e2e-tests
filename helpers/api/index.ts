import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
