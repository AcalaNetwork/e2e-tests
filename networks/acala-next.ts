import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
