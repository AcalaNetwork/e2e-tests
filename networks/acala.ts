import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
