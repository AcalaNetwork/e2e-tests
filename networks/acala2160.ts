import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
