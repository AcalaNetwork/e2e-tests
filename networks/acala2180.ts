import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
