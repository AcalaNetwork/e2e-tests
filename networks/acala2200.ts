import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
