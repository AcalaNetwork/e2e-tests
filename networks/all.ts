import { Config } from './types'

import acala2160Config from './acala2160'
import acalaConfig from './acala'
import hydraDXConfig from './hydraDX'
import moonbeamConfig from './moonbeam'
import polkadot9381Config from './polkadot9381'
import polkadotConfig from './polkadot'
import statemintConfig from './statemint'

const all = {
  polkadot: polkadotConfig,
  polkadot9381: polkadot9381Config,
  statemint: statemintConfig,
  acala: acalaConfig,
  acala2160: acala2160Config,
  moonbeam: moonbeamConfig,
  hydraDX: hydraDXConfig,
} satisfies Record<string, Config>

export default all
