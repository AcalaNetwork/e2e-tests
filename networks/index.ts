import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
