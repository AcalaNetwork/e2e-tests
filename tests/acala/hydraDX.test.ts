import { beforeEach, describe, expect, it } from 'vitest'
import { connectParachains } from '@acala-network/chopsticks'

import { checkEvents, checkHrmp, checkSystemEvents, sendTransaction, testingPairs } from '../helper'
import { xTokens } from '../api/extrinsics'
import networks, { Network } from '../networks'

describe('Acala <-> HydraDX', () => {
  let hydraDX: Network
  let acala: Network

  const { alice } = testingPairs()

  beforeEach(async () => {
    hydraDX = await networks.hydraDX({ blockNumber: 2136151 })
    acala = await networks.acala({ blockNumber: 3143845 })
    await connectParachains([hydraDX.chain, acala.chain])

    await acala.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: 'DOT' }], { free: 10 * 1e12 }],
          [[alice.address, { Token: 'AUSD' }], { free: 10 * 1e12 }],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    })
    await hydraDX.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 1000 * 1e12 } }]],
      },
    })

    return async () => {
      await hydraDX.teardown()
      await acala.teardown()
    }
  })

  it('Acala transfer DAI to hydraDX', async () => {
    const DAI = '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae'
    const tx0 = await sendTransaction(
      acala.api.tx.sudo
        .sudoAs(
          'zvX8kCdZvWwuGdKgePE6tPa4wgJnxiJsvM4CkC4kZ9QE6Hp',
          acala.api.tx.currencies.transfer(alice.address, { Erc20: DAI }, '1000000000000000000')
        )
        .signAsync(alice, { nonce: 0 })
    )

    await acala.chain.newBlock()

    await checkEvents(tx0, 'ClaimAccount', 'currencies').toMatchSnapshot()

    const tx1 = await sendTransaction(
      xTokens(acala.api, false, '2034', { Erc20: DAI }, '1000000000000000000', alice.addressRaw).signAsync(alice, {
        nonce: 1,
      })
    )

    await acala.chain.newBlock()

    await checkEvents(tx1, 'xTokens', 'xcmpQueue', 'evm').toMatchSnapshot()
    await checkHrmp(acala).toMatchSnapshot()

    await hydraDX.chain.newBlock()
    expect((await hydraDX.api.query.tokens.accounts(alice.address, '13')).toHuman()).toMatchInlineSnapshot(`
      {
        "free": "0",
        "frozen": "0",
        "reserved": "0",
      }
    `)
  })

  it('hydraDX tranfer DAI to acala', async () => {
    const DAI = '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae'
    const tx0 = await sendTransaction(
      acala.api.tx.sudo
        .sudoAs(
          'zvX8kCdZvWwuGdKgePE6tPa4wgJnxiJsvM4CkC4kZ9QE6Hp',
          acala.api.tx.currencies.transfer(alice.address, { Erc20: DAI }, '1000000000000000000')
        )
        .signAsync(alice, { nonce: 0 })
    )

    await acala.chain.newBlock()

    await checkEvents(tx0, 'ClaimAccount', 'currencies').toMatchSnapshot()

    const tx1 = await sendTransaction(
      xTokens(acala.api, false, '2034', { Erc20: DAI }, '1000000000000000000', alice.addressRaw).signAsync(alice, {
        nonce: 1,
      })
    )

    await acala.chain.newBlock()

    await checkEvents(tx1, 'xTokens', 'xcmpQueue', 'evm').toMatchSnapshot()
    await checkHrmp(acala).toMatchSnapshot()

    await hydraDX.chain.newBlock()
    expect((await hydraDX.api.query.tokens.accounts(alice.address, '13')).toHuman()).toMatchInlineSnapshot(`
      {
        "free": "0",
        "frozen": "0",
        "reserved": "0",
      }
    `)

    await hydraDX.api.tx.xTokens
      .transfer(
        '13',
        '995600000000000000',
        {
          V1: {
            parents: 1,
            interior: {
              X2: [
                {
                  Parachain: '2000',
                },
                {
                  AccountId32: {
                    network: 'Any',
                    id: alice.addressRaw,
                  },
                },
              ],
            },
          },
        },
        '5000000000'
      )
      .signAndSend(alice)
    await hydraDX.chain.newBlock()
    expect((await hydraDX.api.query.tokens.accounts(alice.address, '13')).toHuman()).toMatchInlineSnapshot(`
      {
        "free": "0",
        "frozen": "0",
        "reserved": "0",
      }
    `)

    await acala.chain.newBlock()
    await checkSystemEvents(acala).toMatchSnapshot()
  })
})
