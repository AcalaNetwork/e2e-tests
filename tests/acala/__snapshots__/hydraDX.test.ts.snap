// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Acala <-> HydraDX > Acala transfer DAI to hydraDX > events 1`] = `
[
  {
    "data": {
      "amount": 1000000000000000000,
      "currencyId": {
        "Erc20": "0x54a37a01cd75b616d63e0ab665bffdb0143c52ae",
      },
      "from": "zvX8kCdZvWwuGdKgePE6tPa4wgJnxiJsvM4CkC4kZ9QE6Hp",
      "to": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "Transferred",
    "section": "currencies",
  },
]
`;

exports[`Acala <-> HydraDX > Acala transfer DAI to hydraDX > events 2`] = `
[
  {
    "data": {
      "contract": "0x54a37a01cd75b616d63e0ab665bffdb0143c52ae",
      "from": "0x4d9cf56ac6f1b4b6be4757ad0af3348fc6a918c2",
      "logs": [
        {
          "address": "0x54a37a01cd75b616d63e0ab665bffdb0143c52ae",
          "data": "(hash)",
          "topics": [
            "(hash)",
            "(hash)",
            "(hash)",
          ],
        },
      ],
      "usedGas": "(rounded 50000)",
      "usedStorage": -64,
    },
    "method": "Executed",
    "section": "evm",
  },
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X2": [
                  {
                    "Parachain": 2000,
                  },
                  {
                    "GeneralKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2000)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Any",
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000000,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 2000,
                },
                {
                  "GeneralKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                },
              ],
            },
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`Acala <-> HydraDX > Acala transfer DAI to hydraDX > hrmp 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2000,
                        },
                        {
                          "generalKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2000,
                        },
                        {
                          "generalKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 2034,
  },
]
`;

exports[`Acala <-> HydraDX > hydraDX tranfer DAI to acala > events 1`] = `
[
  {
    "data": {
      "amount": 1000000000000000000,
      "currencyId": {
        "Erc20": "0x54a37a01cd75b616d63e0ab665bffdb0143c52ae",
      },
      "from": "zvX8kCdZvWwuGdKgePE6tPa4wgJnxiJsvM4CkC4kZ9QE6Hp",
      "to": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "Transferred",
    "section": "currencies",
  },
]
`;

exports[`Acala <-> HydraDX > hydraDX tranfer DAI to acala > events 2`] = `
[
  {
    "data": {
      "contract": "0x54a37a01cd75b616d63e0ab665bffdb0143c52ae",
      "from": "0x4d9cf56ac6f1b4b6be4757ad0af3348fc6a918c2",
      "logs": [
        {
          "address": "0x54a37a01cd75b616d63e0ab665bffdb0143c52ae",
          "data": "(hash)",
          "topics": [
            "(hash)",
            "(hash)",
            "(hash)",
          ],
        },
      ],
      "usedGas": "(rounded 50000)",
      "usedStorage": -64,
    },
    "method": "Executed",
    "section": "evm",
  },
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X2": [
                  {
                    "Parachain": 2000,
                  },
                  {
                    "GeneralKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2000)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Any",
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000000,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 2000,
                },
                {
                  "GeneralKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                },
              ],
            },
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`Acala <-> HydraDX > hydraDX tranfer DAI to acala > hrmp 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2000,
                        },
                        {
                          "generalKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": "0x00000000000000000de0b6b3a7640000",
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "parachain": 2000,
                        },
                        {
                          "generalKey": "0x0254a37a01cd75b616d63e0ab665bffdb0143c52ae",
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 2034,
  },
]
`;

exports[`Acala <-> HydraDX > hydraDX tranfer DAI to acala > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 39000000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 2200000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1200000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 6600000000)",
      "currencyId": {
        "Token": "AUSD",
      },
      "who": "23M5ttkmR6KcnvsNJdmYTpLo9xfc54g8uCk55buDfiJPon69",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "Yes",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 260000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
  {
    "data": {
      "dispatchInfo": {
        "class": "Mandatory",
        "paysFee": "No",
        "weight": {
          "proofSize": 0,
          "refTime": "(rounded 99000000)",
        },
      },
    },
    "method": "ExtrinsicSuccess",
    "section": "system",
  },
]
`;
